misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/expression/QueryExpressionEvaluator.java

```diff
@@ -16,80 +16,31 @@
-import static org.ballerinalang.debugadapter.evaluation.utils.EvaluationUtils.getRuntimeMethod;
 
 /**
  * Query expression evaluator implementation.
  *
  * @since 2.0.0
  */
-public class QueryExpressionEvaluator extends Evaluator {
-
-    private final QueryExpressionNode syntaxNode;
+public class QueryExpressionEvaluator extends ExpressionAsProgramEvaluator {
 
     public QueryExpressionEvaluator(EvaluationContext context, QueryExpressionNode queryExpressionNode) {
-        super(context);
-        this.syntaxNode = queryExpressionNode;
+        super(context, getQueryWithTypeCasts(queryExpressionNode));
     }
 
-    @Override
-    public BExpressionValue evaluate() throws EvaluationException {
-        QueryExpressionProcessor queryProcessor = new QueryExpressionProcessor(context, syntaxNode);
-        try {
-            Map.Entry<Path, String> execPathAndMainClassName = queryProcessor.generateExecutables();
-            Path executablePath = execPathAndMainClassName.getKey();
-            String mainClassName = execPathAndMainClassName.getValue();
-
-            List<String> argTypes = new ArrayList<>();
-            argTypes.add(JAVA_STRING_CLASS);
-            argTypes.add(JAVA_STRING_CLASS);
-            argTypes.add(JAVA_STRING_CLASS);
-            argTypes.add(JAVA_OBJECT_ARRAY_CLASS);
-            RuntimeStaticMethod loadQueryClasses = getRuntimeMethod(context, B_DEBUGGER_RUNTIME_CLASS,
-                    CLASSLOAD_AND_INVOKE_METHOD, argTypes);
-
-            List<Value> argList = new ArrayList<>();
-            argList.add(getAsJString(context, executablePath.toAbsolutePath().toString()));
-            argList.add(getAsJString(context, mainClassName));
-            argList.add(getAsJString(context, QUERY_FUNCTION_NAME));
-
-            // adds all the captured variable values as rest arguments.
-            argList.addAll(queryProcessor.getExternalVariableValues());
-            loadQueryClasses.setArgValues(argList);
-            loadQueryClasses.invokeSafely();
-            Value queryResult = loadQueryClasses.invokeSafely();
-
-            return new BExpressionValue(context, queryResult);
-        } catch (EvaluationException e) {
-            throw e;
-        } catch (Exception e) {
-            throw createEvaluationException(INTERNAL_ERROR, syntaxNode.toSourceCode().trim());
-        } finally {
-            // Need to dispose the query processor, which will clean up temporary resources generated during the
-            // evaluation.
-            queryProcessor.dispose();
+    private static ExpressionNode getQueryWithTypeCasts(QueryExpressionNode syntaxNode) {
+        if (syntaxNode.parent() != null && syntaxNode.parent().kind() == SyntaxKind.TYPE_CAST_EXPRESSION) {
+            // Since Ballerina query expression return type might depend on the contextual type, we need to inject
+            // any existing type castings o top of the query expressions (expression will results in semantic errors
```

Comment:
Is it "any existing type castings `to`" ?

Reply:
it should be "on".. will fix the typo